src/rust/engine/src/core.rs

```diff
@@ -164,22 +164,25 @@ pub const ANY_TYPE: TypeId = TypeId(0);
 pub struct Function(pub Key);
 
 impl Function {
-  fn pretty_print(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+  pub fn name(&self) -> String {
     let Function(key) = self;
+    let module = externs::project_str(&externs::val_for(&key), "__module__");
     let name = externs::project_str(&externs::val_for(&key), "__name__");
-    write!(f, "{}()", name)
+    let code = externs::project_ignoring_type(&externs::val_for(&key), "__code__");
+    let lineno = externs::project_str(&code, "co_firstlineno");
```

Comment:
Very very neat that we can do this complex python-level logic while staying within rust!

Reply:
Yes!! Thanks to @gshuflin for showing me how to do this!